### PR TITLE
[MODULAR] Reagent Forging Chem Blacklist

### PR DESCRIPTION
**Pull request by Nerev4r**
## About The Pull Request
toppng com-filthy-frank-its-time-to-stop-transparent-627x858

                                                     It's time to stop.

This implements a shared blacklist used by reagent weapons and reagent clothing, to prevent blatantly imbalanced combinations of chems and to prevent stuff like creation of the Sacred Cum Blade (sorry!)

## How This Contributes To The Skyrat Roleplay Experience
Reagent forging grows ever-more powerful, and it's about time we took a second look at it. Stuff like pax spears or teslium/enriched liquid electricity spears are essentially a game-ender if you aren't currently dosed with a purging chem, and it's an extremely unpleasant experience to be hit by not only a reagent-based ggnore, but also to be taking quite substantial damage just from the application of it. I don't even think thick clothing stops this stuff.

## Changelog

:cl:
balance: Reagent forging now has a reagent blacklist.
/:cl:

### CODE REVIEW

**Comment by Stalkeros2** [comment]
Nooooo not the breast milk. :(:(:(:(:(

**Comment by RimiNosha** [comment]
Don't forget anything under `/reagent/aphrodisiac`.

File: modular_skyrat/modules/reagent_forging/code/forge.dm [hunk 1]
```diff
@@ -357,6 +400,10 @@
 			if(weapon_reagent.volume < MIN_IMBUE_REQUIRED)
 				attacking_weapon.reagents.remove_all_type(weapon_reagent.type)
 				continue
+			if(is_type_in_list(weapon_reagent, disallowed_reagents))
```
Comment (Zonespace27) [inline modular_skyrat/modules/reagent_forging/code/forge.dm]: make it a typecache instead

**Comment by dawsonkeyes** [comment]
Reagent Forging 3.0: You Can (Not) Imbue

I still think allowing reagent imbuement at all is excessive, but this list is a good halfway between removal and not doing anything at all

**Comment by YakumoChen** [comment]
With a black list like that you may as well remove imbuing entirely

Aphro I'm a bit against, it's a staple for erpers to make imbued collars and such

**Comment by YakumoChen** [comment]
What makes this more novel than syringe guns? You generally have to get up close to hit with these and most of the poisons that don't stun (most of this list in fact) arent significantly more additive in a fight. Beating someone to death with a sword will usually be effective far sooner than most toxins will to do anything notable.

**Comment by Nerev4r** [comment]
> What makes this more novel than syringe guns? You generally have to get up close to hit with these and most of the poisons that don't stun (most of this list in fact) arent significantly more additive in a fight. Beating someone to death with a sword will usually be effective far sooner than most toxins will to do anything notable.

The problem is that the poisons that do/did stun are essentially a few-tap win, especially given that reagent spears do it at a range while having more damage than bone spears (one more damage, but still.) Imbuing should only be additive, it shouldn't be a method of, say, instantly delivering people teslium *while* doing crazy damage to them, and I'm still 99% that this stuff doesn't respect thick clothing or anything at all.

**Comment by YakumoChen** [comment]
Most of this list isn't stun chemicals. Even novel chems like hallucinogens are removed without given reason

If thick clothing is an issue then add a check for it?

**Comment by SpaceVampire** [comment]
You really might as well just remove the system instead of making it useless.
You could also rework it into adding a limited number of custom special effects.
Instead of just any chemical, but not the chemicals on the special blacklist.
I get that is a bit outside of the scope, but when it comes to this via removal.
I would rather have that then a system where there is a bunch of stuff that just doesn't work.
That you can only figure out via trying and failing or looking at the code directly.
Like-it seems like everything you can do with the system is either way too good or utterly useless.
Which very little middle-ground and that isn't really a good stare for the feature.
Especially given the massive length of your currently planned list.
(I also agree that aphrodisiacs are fine to keep if you are going along with the whitelist. Stuff like the cum sword is clearly silly and should be removed, but I have seen people making sexy collars before and I can't recall it leading to any prefbreak issues.)

**Comment by dawsonkeyes** [comment]
> You really might as well just remove the system instead of making it useless.
> Like-it seems like everything you can do with the system is either way too good or utterly useless.

Yes

**Comment by Nerev4r** [comment]
https://www.youtube.com/watch?v=wxL4pzuwSUQ

**Comment by jjpark-kb** [comment]
Aphrodisiacs should not be blacklisted, just as Yakumo said-- its a staple for the lewd collars. If someone uses lewdchems (and they don't have prefs within them or something), then thats a separate issue.  Toxins are purged by the liver fast, so toxins are not really a concern.

Yea, I just don't agree with blacklisting chemicals.

Side note to those whom it concerns: we are not removing imbuing, its staying.

**Comment by YakumoChen** [comment]
Can you explain the reason you've blacklisted:
-numerous non-stun toxins ie cyanide, acid, napalm, etc
-non-damaging chemicals like mutetoxin
-enhancement drugs eg meth
-medicines

**Comment by dawsonkeyes** [comment]
> Side note to those whom it concerns: we are not removing imbuing, its staying.

who is we

**Comment by Nerev4r** [comment]
reintroducing aphrodesiacs but i'd remind that it probably should be bwoinkable to be wearing crocin rings/whatever outside of dorms

**Comment by Nerev4r** [comment]
> Can you explain the reason you've blacklisted: -numerous non-stun toxins ie cyanide, acid, napalm, etc -non-damaging chemicals like mutetoxin -enhancement drugs eg meth -medicines

Acid does extra brute damage (extra hard-hitting if we merge stuff like titanium hammers,) cyanide can stun for 4 whole-ass seconds, napalm metabolizes rather slow and can burn the shit out of someone if you even lightly heat them. Mute toxin is a 'I hit you and you can't call for help at all' for free. Meth is a hilarious speedboost that'd last a long time if you simply devote your hand and neck slots to it. Earthsblood is a pax, morphine is a sleep, ephedrine is meth-lite, synaptizine is an anti-stun, and penthrite/epinephrine can be used to explode miners.

**Comment by dawsonkeyes** [comment]
> Acid does extra brute damage (extra hard-hitting if we merge stuff like titanium hammers,) cyanide can stun for 4 whole-ass seconds, napalm metabolizes rather slow and can burn the shit out of someone if you even lightly heat them. Mute toxin is a 'I hit you and you can't call for help at all' for free. Meth is a hilarious speedboost that'd last a long time if you simply devote your hand and neck slots to it. Earthsblood is a pax, morphine is a sleep, ephedrine is meth-lite, synaptizine is an anti-stun, and penthrite/epinephrine can be used to explode miners.

oh my god i didnt even stop to consider epinephrine ash walker spears versus miners with pens, that's horrendous

**Comment by YakumoChen** [comment]
You have not listed any reasons in the PR body or why it is good for the game why any of these should be removed 

The weapons can be toned down with relative ease. Gutting any interesting chemicals for use in imbuing like this removes any potential the chems have. Especially with the fairly high requirements required to get an imbued weapon in the first place.

Chems like mute toxin already see little to no use in practical gameplay and I fail to see how this severely detracts from combat by keeping it in.

**Comment by YakumoChen** [comment]
Even chems like morphine I don't see a reason to remove. Morphine sleep is severely delayed, requires high dosage, and even removes slowdown before then (it buffs the victim on application)

**Comment by dawsonkeyes** [comment]
i think certain chemicals such as pax, teslium, and enriched liquid electricity and chems with similar nigh-immediate debilitating effects that render people unable to fight are what the list should include, with slow-burn stuff like morphine and even mute toxin not being as terrible to fight by comparison

**Comment by YakumoChen** [comment]
Some of these choices feel alarmist and not based on actual round play. Have ash lizards ever actually exploded a miner with ephine? Do ash lizards even get practical access to chem dispensers? Injected napalm burns you but how do you heat it mid-flight? Do we blacklist water and potassium because someone could dual wield? Can you make a meth ring that doesn't just kill you with brain damage? How is that different from having a pill?

**Comment by Nerev4r** [comment]
> i think certain chemicals such as pax, teslium, and enriched liquid electricity and chems with similar nigh-immediate debilitating effects that render people unable to fight are what the list should include, with slow-burn stuff like morphine and even mute toxin not being as terrible to fight by comparison

Cut mute toxin/morphine.

**Comment by YakumoChen** [comment]
I think he meant "blacklist all the ones that are immediately debilitating in combat (pax, teslium, electricity) " and purge the rest of the list.

If problem chems are made obvious in documented gameplay the list can always be added to.

Traitor only chems shouldn't be added, their usage is appropriate and they are already limited availability

**Comment by Nerev4r** [comment]
The list is now var/static/list/disallowed_reagents = typecacheof(list(
		/datum/reagent/inverse/,
		/datum/reagent/consumable/entpoly,
		/datum/reagent/pax,
		/datum/reagent/consumable/liquidelectricity/enriched,
		/datum/reagent/toxin/heparin,
		/datum/reagent/teslium,
		/datum/reagent/eigenstate,
		/datum/reagent/drug/bath_salts,
		/datum/reagent/drug/methamphetamine,
		/datum/reagent/drug/pcp,
		/datum/reagent/consumable/cum,
		/datum/reagent/consumable/femcum,
		/datum/reagent/consumable/breast_milk,
		/datum/reagent/toxin/chloralhydrate,
		/datum/reagent/toxin/bonehurtingjuice,
		/datum/reagent/toxin/lexorin,
		/datum/reagent/toxin/staminatoxin,
		/datum/reagent/toxin/sulfonal,
		/datum/reagent/toxin/histamine,
		/datum/reagent/medicine/atropine,
		/datum/reagent/toxin/fentanyl,
		/datum/reagent/toxin/acid,
		/datum/reagent/phlogiston,
		/datum/reagent/napalm,
		/datum/reagent/thermite,
		/datum/reagent/medicine/earthsblood,
		/datum/reagent/medicine/ephedrine,
		/datum/reagent/medicine/epinephrine,
		/datum/reagent/medicine/synaptizine,
		/datum/reagent/medicine/c2/penthrite,
	))
	
Everything that pacifies, stuns, does brute/burn upfront, sleeps, explodes, lights people on fire, does a bunch of stamina, or has weird unpredictable side effects like inverse chems, has been kept on the list. As well as buff chems like PCP.

**Comment by dawsonkeyes** [comment]
> I think he meant "blacklist all the ones that are immediately debilitating in combat (pax, teslium, electricity) " and purge the rest of the list.
> 
> If problem chems are made obvious in documented gameplay the list can always be added to.
> 
> Traitor only chems shouldn't be added, their usage is appropriate and they are already limited availability

traitor only chems like spewium aren't obtainable in volumes of 100 units or more without purchasing multiple chemical kits, so yea

**Comment by YakumoChen** [comment]
Remove the buff chems, its functionally the same as carrying pills
Remove the "stamina damage only chems, this is not additive to lethal combat, plus stun weapons exist already
Damage / toxic chems are not a notable problem, livers exist, blacklisting them hardly makes sense ICly either.
Lexorin can be purged with an epipen

The list should be limited to immediate pacification or stun threats that are easily obtainable (not antag chems)

File: modular_skyrat/modules/reagent_forging/code/forge.dm [hunk 2]
```diff
@@ -78,6 +78,40 @@
...
+		/datum/reagent/consumable/breast_milk,
+		/datum/reagent/toxin/chloralhydrate,
+		/datum/reagent/toxin/bonehurtingjuice,
+		/datum/reagent/toxin/lexorin,
+		/datum/reagent/toxin/staminatoxin,
+		/datum/reagent/toxin/sulfonal,
+		/datum/reagent/toxin/histamine,
+		/datum/reagent/medicine/atropine,
+		/datum/reagent/toxin/fentanyl,
+		/datum/reagent/toxin/acid,
```
Comment (jjpark-kb) [inline modular_skyrat/modules/reagent_forging/code/forge.dm]: they get filtered from the liver, remove these.

File: modular_skyrat/modules/reagent_forging/code/forge.dm [hunk 3]
```diff
@@ -78,6 +78,40 @@
...
+		/datum/reagent/phlogiston,
+		/datum/reagent/napalm,
+		/datum/reagent/thermite,
+		/datum/reagent/medicine/earthsblood,
+		/datum/reagent/medicine/ephedrine,
+		/datum/reagent/medicine/epinephrine,
+		/datum/reagent/medicine/synaptizine,
+		/datum/reagent/medicine/c2/penthrite,
```
Comment (jjpark-kb) [inline modular_skyrat/modules/reagent_forging/code/forge.dm]: remove these-- medicines allow people to create heal staves and other misc. items.

File: modular_skyrat/modules/reagent_forging/code/forge.dm [hunk 4]
```diff
@@ -78,6 +78,40 @@
...
+		/datum/reagent/inverse/,
+		/datum/reagent/consumable/entpoly,
+		/datum/reagent/pax,
+		/datum/reagent/consumable/liquidelectricity/enriched,
+		/datum/reagent/toxin/heparin,
+		/datum/reagent/teslium,
+		/datum/reagent/eigenstate,
+		/datum/reagent/drug/bath_salts,
+		/datum/reagent/drug/methamphetamine,
+		/datum/reagent/drug/pcp,
```
Comment (jjpark-kb) [inline modular_skyrat/modules/reagent_forging/code/forge.dm]: if people make it and get enough of this, they should be allowed to infuse it-- remove these.

File: modular_skyrat/modules/reagent_forging/code/forge.dm [hunk 5]
```diff
@@ -78,6 +78,40 @@
 	///list containing cooking radial buttons, for when anything of /obj/item/food is used on the forge
 	var/static/list/radial_options = list("oven" = radial_oven, "microwave" = radial_microwave)
 
+    ///Blacklist that contains reagents that weapons and armor are unable to be imbued with.
+	var/static/list/disallowed_reagents = typecacheof(list(
+		/datum/reagent/inverse/,
+		/datum/reagent/consumable/entpoly,
+		/datum/reagent/pax,
+		/datum/reagent/consumable/liquidelectricity/enriched,
+		/datum/reagent/toxin/heparin,
```
Comment (jjpark-kb) [inline modular_skyrat/modules/reagent_forging/code/forge.dm]: toxins purged, remove.

**Review by jjpark-kb — CHANGES_REQUESTED** [review]
I have made requested changes.

**Comment by Nerev4r** [comment]
var/static/list/disallowed_reagents = typecacheof(list(
        /datum/reagent/inverse/,
        /datum/reagent/consumable/entpoly,
        /datum/reagent/pax,
        /datum/reagent/consumable/liquidelectricity/enriched,
        /datum/reagent/teslium,
        /datum/reagent/eigenstate,
        /datum/reagent/drug/pcp,
        /datum/reagent/consumable/cum,
        /datum/reagent/consumable/femcum,
        /datum/reagent/consumable/breast_milk,
        /datum/reagent/toxin/acid,
        /datum/reagent/phlogiston,
        /datum/reagent/napalm,
        /datum/reagent/thermite,
        /datum/reagent/medicine/earthsblood,
        /datum/reagent/medicine/ephedrine,
        /datum/reagent/medicine/epinephrine,
    ))
    
   This okay?

**Comment by jjpark-kb** [comment]
yes, way better. @YakumoChen thoughts?

**Comment by YakumoChen** [comment]
You have not removed all items from the list that @jjpark-kb requested, explain?

Pax and Teslium/Liquid Electricity are the only three major ones I see worth keeping on the list that were requested to be removed, so I'm fine with those three remaining myself.

Depending on the dosage though that reagent weapons inject, the liver might outright remove them before they tick.

**Comment by Nerev4r** [comment]
For posterity, items were not removed due to discussion in #development-code.

**Comment by Nerev4r** [comment]
image
image
we vibin'

**Pull request merged**